pkg/rhelv2/rpm/rpm.go

```diff
@@ -57,24 +59,117 @@ const (
+}
+
 // AllRHELRequiredFiles lists all the names of the files required to identify RHEL-based releases.
 var AllRHELRequiredFiles set.StringSet
 
 func init() {
-	AllRHELRequiredFiles.Add(dbPath)
 	AllRHELRequiredFiles.AddAll(RequiredFilenames()...)
 	AllRHELRequiredFiles.AddAll(redhatrelease.RequiredFilenames...)
 	AllRHELRequiredFiles.AddAll(osrelease.RequiredFilenames...)
 }
 
+// rpmDatabase represents an RPM database in the filesystem.
+type rpmDatabase struct {
+	// The path to the rpm database, can be used as --dbpath <dbPath> in rpm commands.
+	dbPath string
+}
+
+// createRPMDatabaseFromImage creates an RPM database in a temporary directory
+// from the RPM database found in the container image. All known RPM database
+// backend is supported (i.e. bdb, sqlite). If no database is found in the image,
+// returns nil.
+func createRPMDatabaseFromImage(imageFiles tarutil.LayerFiles) (*rpmDatabase, error) {
+	// Find all known RPM database files in the image archive.
+	dbFiles := make(map[string]tarutil.FileData)
+	for _, name := range rpmDatabaseFiles {
+		data, exists := imageFiles.Get(path.Join(rpmDatabaseDir, name))
+		if exists {
+			dbFiles[name] = data
+		}
+	}
+	if len(dbFiles) == 0 {
+		// Not rpm database was found.
+		return nil, nil
+	}
+	// Write the database files to the filesystem.
+	dbDir, err := os.MkdirTemp("", "rpm")
+	if err != nil {
+		log.WithError(err).Error("could not create temporary folder for the rpm database")
+		return nil, commonerr.ErrFilesystem
+	}
+	defer func() {
+		// Remove temporary directory if returning on errors.
+		if err != nil {
+			_ = os.RemoveAll(dbDir)
+		}
+	}()
+	for name, data := range dbFiles {
+		path := filepath.Join(dbDir, name)
+		err = os.WriteFile(path, data.Contents, 0700)
+		if err != nil {
+			log.WithError(err).Error("failed to create rpm database file")
+			return nil, commonerr.ErrFilesystem
+		}
+	}
+	return &rpmDatabase{
+		dbPath: dbDir,
+	}, nil
+}
+
+// providesFile return true if a package provides the specified path in the RPM
+// database. If the path is relative, we assume its relative to the root
+// directory.
+func (d *rpmDatabase) providesFile(path string) bool {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`-q`,
+		`--whatprovides`, path,
+	)
+	if err := cmd.Run(); err != nil {
+		// When rpm does not provide a file, the expected exit status is 1. On non-zero
+		// we always default to returning the file is not provided, but anything other
+		// than 1 is considered unexpected.
+		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() != 1 {
+			log.WithError(err).Errorf(
+				"unexpected exit status when querying %s is provided by an RPM package", path)
+		}
+		return false
+	}
+	// The command exited properly, which implies the file IS provided by an RPM
+	// package.
+	return true
+}
+
+// delete removes all the RPM database files.
+func (d *rpmDatabase) delete() error {
```

Comment:
This is pretty much the same semantics as `Close() error`. I think we should rename this to `Close()` so it is clear `rpmDatabase` is an `io.Closer`, which should be closed upon usage

Reply:
Why do we want to make an RPM database an `io.Closer`? To me, `io.Closer` communicates the RPM database object was once opened for I/O, and now is going to be closed for reading/writing, but not deleted. However, `delete()` will remove the database altogether and all its data -- by the way, I think I/O in the `io` package is about file I/O, not querying or writing to databases, but anyway.

The two functions have the same signature, and they are both called when the object is no longer used, but they have different semantics, and for that, they are two different types of functions, hence, a RPM DB is not an io.Closer.

Reply:
thinking about it more, and I agree making this a `Closer` simply because this matches the signature doesn't make sense. Would make sense if this wrapped around some other `Close` method or this were really a `ReadCloser` or something

Reply:
I'd argue that having other `Close()` methods in its implementation would still not be a good reason per se to rename `delete()` to `Close()`. It would make sense if the methods had the same semantics as you initially pointed out.